templates/_macros/hero.jinja

```diff
@@ -0,0 +1,109 @@
+{% from "_macros/cta/cta-block.jinja" import cta_block %}
+
+# Params
+# title_text: Hero title text (required)
+# subtitle_text: Hero subtitle text
+# is_full_width_on_tablet: whether image is in its own row on tablet.
+#   If false, image will be in its own row (stacked below header/description)
+# image_src: image source URL (required)
+# image_alt: image alt text
+# image_aspect_ratio: aspect ratio identifier of the image, from the `p-image-container` component
+# image_full_cover: whether the image should span the full width of the bottom of the hero section
+# image_hidden_breakpoints: list of breakpoints where the image should be hidden, e.g. ["large", "medium"] to only show on small screens
+# show_cta: whether to show the call-to-action block
+# cta_primary_text: primary CTA button text, shown in a primary-colored button
+# cta_primary_url: primary CTA button URL
+# cta_secondary_text: secondary CTA button text, shown as a link
+# cta_secondary_url: secondary CTA button URL
+# Slots
+# description: slot for additional paragraph-style content below the title and subtitle
+
+{% macro hero(
+  title_text="",
+  subtitle_text="",
+  is_full_width_on_tablet = true,
+  image_src = "",
+  image_alt = "",
+  image_aspect_ratio = "",
+  image_full_cover = false,
+  image_hidden_breakpoints = [],
+  show_cta = false,
+  cta_primary_text = "Learn more",
+  cta_primary_url = "#",
+  cta_secondary_text = "",
+  cta_secondary_url = "#"
+)-%}
+  {% set has_subtitle = subtitle_text|trim|length > 0 %}
+  {% set description_content = caller('description') %}
+  {% set has_description = description_content|trim|length > 0 %}
+  {% set has_specific_image_aspect_ratio = image_aspect_ratio|trim|length > 0 %}
+  {% set has_cta = show_cta and cta_primary_text|trim|length > 0 %}
+
+  {% macro hero_title_block() -%}
```

Comment:
My thinking was that there is some content that has to live at different places in the markup depending on variants (flags) - so rather than define them and their surrounding containers etc multiple times it would be nice to simply define them after all of the macro-scoped variables etc, so the "internal" macros have access to them without needing to worry about passing them around as parameters.

This makes things for the calling portion of the macro quite convenient, and makes the main body of the macro very readable at a high level:

```html
<div class="col">
  {% if image_full_cover %}
    {{ hero_title_block() }}
  {% else %}
    <div class="p-section--shallow">
      {{ hero_title_block() }}
      {{ hero_subtitle_block() }}
    </div>
    {{ hero_description_block() }}
    {{ hero_cta_block() }}
  {% endif %}
</div>
```